core/validatorapi/validatorapi.go

```diff
@@ -666,6 +671,93 @@ func (c Component) SubmitBeaconCommitteeSubscriptionsV2(ctx context.Context, sub
 	return c.getCommResponse(ctx, psigsBySlot)
 }
 
+// AggregateAttestation fetches the aggregate attestation given an attestation.
+// Note: It queries aggregated attestation from DutyDB (this is blocking).
+func (c Component) AggregateAttestation(ctx context.Context, slot eth2p0.Slot, attestationDataRoot eth2p0.Root) (*eth2p0.Attestation, error) {
+	return c.awaitAggAttFunc(ctx, int64(slot), attestationDataRoot)
+}
+
+// SubmitAggregateAttestations receives partially signed aggregateAndProofs.
+func (c Component) SubmitAggregateAttestations(ctx context.Context, aggregateAndProofs []*eth2p0.SignedAggregateAndProof) error {
+	var valIdxs []eth2p0.ValidatorIndex
+	for _, agg := range aggregateAndProofs {
+		valIdxs = append(valIdxs, agg.Message.AggregatorIndex)
+	}
+
+	vals, err := c.eth2Cl.Validators(ctx, "head", valIdxs)
+	if err != nil {
+		return err
+	}
+
+	// Get selection proofs (slot signatures).
+	for _, agg := range aggregateAndProofs {
+		duty := core.NewPrepareAggregatorDuty(int64(agg.Message.Aggregate.Data.Slot))
+		eth2Pubkey, err := vals[agg.Message.AggregatorIndex].PubKey(ctx)
+		if err != nil {
+			return err
+		}
+
+		pk, err := core.PubKeyFromBytes(eth2Pubkey[:])
+		if err != nil {
+			return err
+		}
+
+		// Query aggregated subscription from aggsigdb for each duty and public key (this is blocking).
+		s, err := c.aggSigDBFunc(ctx, duty, pk)
+		if err != nil {
+			return err
+		}
+
+		sub, ok := s.(core.SignedBeaconCommitteeSubscription)
+		if !ok {
+			return errors.New("slot signature not found")
+		}
+
+		// Override selection proof with slot signature from subscription.
+		agg.Message.SelectionProof = sub.SlotSignature
+	}
+
+	psigsBySlot := make(map[eth2p0.Slot]core.ParSignedDataSet)
+	for _, agg := range aggregateAndProofs {
+		eth2Pubkey, err := vals[agg.Message.AggregatorIndex].PubKey(ctx)
+		if err != nil {
+			return err
+		}
+
+		pk, err := core.PubKeyFromBytes(eth2Pubkey[:])
+		if err != nil {
+			return err
+		}
+
+		// Verify Signature.
+		err = c.verifyPartialSig(pk, func(pubshare *bls_sig.PublicKey) error {
+			return signing.VerifyAggregateAndProof(ctx, c.eth2Cl, pubshare, agg)
+		})
+		if err != nil {
+			return err
+		}
+
+		_, ok := psigsBySlot[agg.Message.Aggregate.Data.Slot]
```

Comment:
I'd define a `slot := psigsBySlot[agg.Message.Aggregate.Data.Slot` variable to be less verbose